### PR TITLE
Sugar Ant - 2FA and IVAs (GSI-555)

### PR DESCRIPTION
**Pull request by Cito**
Technical specification for the implementation of 2FA and IVAs in the frontend and in the backend.

### CODE REVIEW

File: 46-sugar-ant/technical_specification.md [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+The `ghga-service-commons` library must be changed to reflect the changes in the internal auth token as outlined above.
+
+### Backend Models
+
+A new `IVA` (independent verification address) model must be added to the User Management service:
+	- `id`: string (unique internal id)
+	- `user_id`: string (internal id of the user)
+	- `type`: enum (phone, fax, postal address, in-person)
+	- `value`: string (the actual phone number)
+	- `verification_code`:  string (random number)
```
Comment (KerstenBreuer) [inline 46-sugar-ant/technical_specification.md:195]: As hash

File: 46-sugar-ant/technical_specification.md [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+
+The `TOTPToken`s will be managed by the Auth Adapter, as explained above.
+
+The `AccessRequest` model must be extended to also include an optional `iva_id` field referencing an IVA that is used to verify the corresponding access grant. This field will be initially set to `None`, and it will be set after access has been allowed.
+
+### Login Flow in the Frontend
+
+The frontend keeps the state of the current user in the session storage. The in the frontend can be one of the following stages:
+- `unauthenticated`
+- `logged-in` (with one factor)
```
Comment (KerstenBreuer) [inline 46-sugar-ant/technical_specification.md:212]: Maybe name differently

**Review by KerstenBreuer — DISMISSED** [review]

File: 46-sugar-ant/technical_specification.md [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+The implementation only supports a single TOTP token per user. If an activated token (a TOTP token that has already been successfully validated at least once) already exists and the `force` flag is set to `true`, then the existing TOTP token will be replaced by the newly created one.
+
+As a side effect of this endpoint, all existing IVAs associated with this user must be deleted, as required in the white paper for 2FA and IVAs.
+
+The QR code should be created in the frontend, e.g. using `react-qr-code` or the `qr-code` web component. If during implementation there are any issues with this approach, it can alternatively also be created in the backend, e.g. using the `segno` library.
+
+- `GET /totp-token` - *checks the existence of a TOTP token*
```
[the remark this answers is not in the document]
Comment (Cito) [reply 46-sugar-ant/technical_specification.md]: Changed now

File: 46-sugar-ant/technical_specification.md [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+This endpoint first verifies that the user has a valid auth session, i.e. has been successfully logged in via LS Login and is a registered user. If this is not the case, it returns the HTTP status `401 Unauthorized`. It then checks whether the user has already created an active TOTP token. If this is not the case, the HTTP status code `404 Not Found` is returned. Otherwise, an empty response with the HTTPS status `204 No Content` is returned.
+
+- `POST /verify-totp` - *verifies a one-time password*
```
[the remark this answers is not in the document]
Comment (Cito) [reply 46-sugar-ant/technical_specification.md]: Changed now

File: 46-sugar-ant/technical_specification.md [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+	- request body:
+		- `user_id`: string (the registered User ID)
+		- `totp`: string (the one-time-password)
+
+This endpoint first verifies that the user has a valid auth session, i.e. has been successfully logged in via LS Login. It then verifies that the session refers to an already registered user, and that the user has the same user ID as specified in the request body. Next, it verifies that this user has already created a TOTP token. Finally, it verifies the given one-time password in `totp` using the current time and a configurable time window. If all verification steps succeed, the token is activated, and the HTTP status `204 No Content` is send in an empty response. Otherwise, if the one-time password could be verified, it responds with the HTTP status `401 Unauthorized`.
+
+- `POST /logout` - *removes the user session*
```
[the remark this answers is not in the document]
Comment (Cito) [reply 46-sugar-ant/technical_specification.md]: Changed now

File: 46-sugar-ant/technical_specification.md [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+	- if not a data steward, the current state must be either `unverified` or `pending`
+	- request body:
+		- `verification_code`:  optional string (random number)
+			- must be set and match the currently stored value when the state is set to `verified`
+			- must be null otherwise
+		- `state`: enum (requested, verified)
+			- must be `requested` when the current state is `unverified`
+			- must be `verified` when the current state is `pending`
+	- response body:
+		- `state`: enum (unverified, requested, pending, verified)
```
[the remark this answers is not in the document]
Comment (Cito) [reply 46-sugar-ant/technical_specification.md]: Split into 5 different RPC-style endpoints now.

File: 46-sugar-ant/technical_specification.md (hunk 1 above)
Comment (Cito) [reply 46-sugar-ant/technical_specification.md]: Changed to only store the hash of the verification code. This made adding another state necessary though.

File: 46-sugar-ant/technical_specification.md (hunk 2 above)
Comment (Cito) [reply 46-sugar-ant/technical_specification.md]: Ok, renamed to not cause confusion with different naming in whitepaper.

File: 46-sugar-ant/technical_specification.md [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+
+A new `TOTPToken` model must be created that includes all attributes required to operate a TOTP token, like the OTP secret key that it is based upon (details will be defined during implementation).
+
+The `TOTPToken`s will be managed by the Auth Adapter, as explained above.
+
+The `AccessRequest` model must be extended to also include an optional `iva_id` field referencing an IVA that is used to verify the corresponding access grant. This field will be initially set to `None`, and it will be set after access has been allowed.
+
+### Login Flow in the Frontend
+
+The frontend keeps the state of the current user in the session storage. The in the frontend can be one of the following stages:
```
Comment (ac-jorellanaf) [inline 46-sugar-ant/technical_specification.md:266]:
```suggestion
The frontend keeps the state of the current user in the session storage. The state in the frontend can be one of the following stages:
```

**Review by lkuchenb — CHANGES_REQUESTED** [review]
Thanks Christoph, very well documented! Great material also to prep the integration team already for the follow-up testbed work.

Minor question / term ambiguity: Are "session ID" and "session token" the same thing?

File: 46-sugar-ant/technical_specification.md [hunk 8]
```diff
@@ -0,0 +1,398 @@
...
+
+The Auth Adapter must be extended with a cookie based auth session management.
+
+This means the Auth Adapter should track user sessions using unique session IDs submitted via cookies. It also maintains an internal session store which stores the current user information and authentication state of the user. The session store should not be used for storing any other information about the state of the application.
+
+The store will be implemented in memory in the first implementation. Another type of cache could be used later that would allow retaining sessions when restarting the auth adapter and support multiple instances of the auth adapter. Therefore the storage mechanism for the session cache should be made easily changeable.
+
+The Auth Adapter creates an auth session and tracks users as soon as they have logged in via LS Login, but not earlier.
+
+Instead of converting the OIDC access token to our internal auth token, the Auth Adapter should now convert the content of the user session into the internal auth token. The internal auth token will change a bit, as outlined in a section below.
```
[the remark this answers is not in the document]
Comment (Cito) [reply 46-sugar-ant/technical_specification.md:51]: The auth session holds all necessary information on the authenticated user (e.g. session ID, internal and external user id, name, email, title, stage of authentication).

- a part of this information (id, name, email, title) is passed in the internal auth token to the micro services
- the session ID (and only this) is passed in the auth cookie to the client to re-identify the client on the next visit

File: 46-sugar-ant/technical_specification.md [hunk 9]
```diff
@@ -0,0 +1,398 @@
...
+
+The Auth Adapter must also be extended with a mechanism that prevents "session riding" attacks, using the "Cookie-to-header token" method. Thereby, the first request that creates the session responds not only with the session cookie, but also with a "CSRF cookie" which must be a unique and unpredictable string, either as a random string or derived from the session token via HMAC and an application secret. Contrary to the session cookie, the CSRF cookie must not be a Http-only token, because it must be read by JavaScript running in the frontend and copied to a CSRF-Token for each request to the backend. The Auth Adapter needs to compare the CSRF Token with the unique string set in the CSRF cookie.
```
[the remark this answers is not in the document]
Comment (Cito) [reply 46-sugar-ant/technical_specification.md]: Actually the flag is called "HttpOnly". I have reworded that paragraph a bit in 037de066.

File: 46-sugar-ant/technical_specification.md [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+    - `text`: string (the secret as text)
+    - `svg`: string (URI as QR-code in SVG format)
+
+This endpoint first verifies that the user has a valid auth session, i.e. has been successfully logged in via LS Login. It then verifies that the session refers to an already registered user, and that the user has the same user ID as specified in the request body. Next, if `force` is not set to `true`, it verifies that this user does not already have an active TOTP token. If any of these verification steps fail, it responds with the HTTP status `401 Unauthorized`. Otherwise, creates a TOTP token and returns its provisioning URI which also contains the secret (seed) used by this token as a query parameter, using the HTTP status `201 Created`.
+
+The implementation only supports a single TOTP token per user. If an activated token (a TOTP token that has already been successfully validated at least once) already exists and the `force` flag is set to `true`, then the existing TOTP token will be replaced by the newly created one.
+
+As a side effect of this endpoint, all existing IVAs associated with this user must be deleted, as required in the white paper for 2FA and IVAs.
```
[the remark this answers is not in the document]
Comment (Cito) [reply 46-sugar-ant/technical_specification.md]: Yes. Thanks for noticing, I have fixed this now.

File: 46-sugar-ant/technical_specification.md [hunk 11]
```diff
@@ -0,0 +1,398 @@
...
+  - request body:
+    - `type`: enum (phone, fax, postal address, in-person)
+    - `value`: string (the actual phone number)
+  - response body:
+    - `id`: string (the ID of the newly created IVA)
+- `DELETE /users/{user-id}/ivas/{iva_id}`
+  - *deletes an existing IVA of the specified user*
+  - auth header: internal token (of data steward or same user)
+
+Additionally, the following three RPC-style will be added:
```
[the remark this answers is not in the document]
Comment (Cito) [reply 46-sugar-ant/technical_specification.md:131]: "RPC-style endpoints" I wanted to say. Fixed now.

File: 46-sugar-ant/technical_specification.md [hunk 12]
```diff
@@ -0,0 +1,398 @@
...
+
+The frontend then requests the user data from the user management service using the `GET /users/{ext_id}` endpoint, passing the LS Login ID and the access token. On the backend side, this will create an auth session as side effect.
+
+If the user is already registered, the frontend gets the internal user ID and the user info (particularly, the name and email) in the response, and the state is set to `registered`.
+
+If the user is not yet registered, the state is set to `needs-registration` and the frontend asks the user to register.
+
+If the user info from LS Login does not match the registered user info, the user will not be considered valid by the backend, the state is set to `needs-reregistration`, and the frontend should show a message accordingly, asking the user to confirm and thereby re-register.
+
+TODO: Should we send a notification to a data steward in this case, or maybe even invalidate existing TOTP tokens and IVAs?
```
[the remark this answers is not in the document]
Comment (Cito) [reply 46-sugar-ant/technical_specification.md:293]: This happens when the name or the email provided by LS login does not match the name and email that we have registered. For instance, the name may be changed by the home organization or the user may have changed their contact email address in LS Login.

File: 46-sugar-ant/technical_specification.md [hunk 13]
```diff
@@ -0,0 +1,398 @@
...
+
+Therefore, the frontend checks whether the user already created a TOTP token using the `OPTIONS /rpc/verify/totp-token` endpoint. If this returns "POST" as allowed method, the user is moved to the `has-totp-token` state, otherwise to the `needs-totp-token` state.
+
+Let's assume the user is in the `needs-totp-token` or `lost-totp-token` state. The frontend then uses the `POST /totp-token` endpoint to create a provisioning URI, whereby the `force` flag should be set if and only if the user is in the state `lost-totp-token`. The frontend then presents the returned URI in form of a QR code to the user and asks the user to scan the QR code using an authenticator app. It should also show a button or link to display the secret as text as fallback for manually entering the secret.
+
+The frontend should recommend using Aegis as authenticator app. Google authenticator is not recommended, since it does not require unlocking the phone, and Authy is not recommend since it stores the secrets in the cloud, and does not provide a means for the user to retrieve them, which makes it impossible to migrate them to another app.
```
[the remark this answers is not in the document]
Comment (Cito) [reply 46-sugar-ant/technical_specification.md:303]: Google and Microsoft Authenticator are also available for iOS. But you're right, Aegis is only available on Android. Will recommend something better than Google and MS for iOS as well.

[the remark this answers is not in the document]
Comment (Cito) [reply 46-sugar-ant/technical_specification.md]: Recommended 2FAS now.

File: 46-sugar-ant/technical_specification.md [hunk 14]
```diff
@@ -0,0 +1,398 @@
...
+    - `id`: string (the ID of the newly created IVA)
+- `DELETE /users/{user-id}/ivas/{iva_id}`
+  - *deletes an existing IVA of the specified user*
+  - auth header: internal token (of data steward or same user)
+
+Additionally, the following three RPC-style will be added:
+
+- `POST /rpc/ivas/{iva_id}/unverify`
+  - *invalidate the specified IVA*
+  - auth header: internal token of a data steward
```
[the remark this answers is not in the document]
Comment (Cito) [reply 46-sugar-ant/technical_specification.md:135]: No, this is just an additional means to quickly revoke the verification by a data steward.

It corresponds to the "invalidate" button in the verification address browser wireframe below.

**Review by lkuchenb — APPROVED** [review]
Thanks!

**Review by KerstenBreuer — APPROVED** [review]
Looks very good, no objections

**Review by ac-jorellanaf — APPROVED** [review]

**Pull request merged**